### PR TITLE
Monster collection view

### LINKED ISSUE #147: Monster Family View
Hello. I have been maintaining a spreadsheet that is valuable in organizing my unit collection.

image

It has rows for each family and columns for each element. Black means that I don't have the unit in question (awakened or locked). Green means the unit is awakened or locked and fully skilled (ie. done). The numbers are the number of skill ups that I still need for each unit. Another column for skill ups in possession would be great (unlocked and unawaked units in family).

This is useful for people trying to prepare for new 2a's and for those of us who want to develop a full collection of units.

I would like to contribute to your project and will create a pull request to work on this. I'm a web developer with no Django / Python experience, however, I'll give it shot. I started on my own app to produce this from the swex json file, but it really belongs in swarfarm, if you would have it.

Thanks for the great work all this time. Without you and Xzandro, this game would be unplayable.

### LINKED ISSUE #176: Possible Skillups doesn't consider Devilmons in Material Storage
Whenever you check for possible skillups, it doesn't consider devilmons from Material Storage
image
image

### PR DESCRIPTION
**Pull request by QuatZo**
I'm creating this PR mostly for backend review now. Not sure if there's other, better and/or faster way of fetching and preparing that data.

Right now, I'm fetching all bestiary Monster instances (limiting query to fields I need: `name`, `com2us_id`, `element`, `skill_group_id` (newly created field in Monster model) and `skill_ups_to_max`. I'm using only `Monster name` and `Natural stars` filters. Then, by iterating over this list (using *itertools* for optimization) I'm preparing records for every `skill_group<->element` pair.

After preparing the data, I'm iterating over all awakened Profile monsters (using ALL filters, excluding material monsters) filling up previously prepared records.

Then, there's 3rd iteration over all unawakened Profile monsters, to get all possible skillups by checking the length of every groupped skill group

Anyone see any better solution for this? The Monster Collection view is kind of tricky, because some records should be visible even if Summoner doesn't have any member of given family. Unless we reduce it so user can see only skill groups which monsters user has, i.e. *don't show Dragon family if user doesn't have any Dragon MonsterInstance.*

P.S. IDE formatted whole CSS file after adding some stuff, which I didn't notice earlier 👎 

resolves #147 
resolves #176

### CODE REVIEW

**Comment by PeteAndersen** [comment]
I ran into a couple issues testing:
1. When loading without submitting a form (i.e. a GET request), `form.cleaned_data` is accessed, which is invalid. This can only be accessed if `form.is_valid()` is true.
2. Got a key error on 'light' element when processing Varis, the light low elemental, in my monster collection. This guy is only available awakened and must have gotten excluded from the `base_monsters` queryset. 

```
Internal Server Error: /profile/porksmash/monster/inventory/
Traceback (most recent call last):
  File "...\swarfarm\venv\lib\site-packages\django\core\handlers\exception.py", line 34, in inner
    response = get_response(request)
  File "...\swarfarm\venv\lib\site-packages\django\core\handlers\base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "...\swarfarm\venv\lib\site-packages\django\core\handlers\base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "...\swarfarm\herders\decorators.py", line 18, in wrap
    return function(request, *args, **kwargs)
  File "...\swarfarm\herders\views\monsters.py", line 165, in monster_inventory
    data = monster_stable[mon.monster.skill_group_id]['elements'][mon.monster.element]
KeyError: 'light'
```

**Comment by QuatZo** [comment]
Should be fixed

File: herders/templates/herders/profile/monster_inventory/collection.html [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+        </div>
+        <div class="panel-body">
+            {% if monster_stable %}
+                <div class="container">
+                    <div class="row">
+                        <div class="col-12 col-sm-10">
+                            <table class="table table-responsive">
+                                <thead>
+                                    <th scope="col"></th>
```
Comment (PeteAndersen) [inline herders/templates/herders/profile/monster_inventory/collection.html]: `<th>` elements should be in a `<tr>`

File: herders/templates/herders/profile/monster_inventory/collection.html [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                                            <td>{% if mon.possible_skillups %}{{ mon.possible_skillups }}{% endif %}</td>
+                                    </tr>
+                                    {% endfor %}
+                                </tbody>
+                            </table>
+                        </div>
+                        <div class="col-12 col-sm-2">
+                            <table class="table table-responsive">
+                                <thead>
+                                    <th scope="col">Legend</th>
```
Comment (PeteAndersen) [inline herders/templates/herders/profile/monster_inventory/collection.html]: same here, `<th>` elements should be in a `<tr>`

**Comment by PeteAndersen** [comment]
I slightly rearranged to put the legend on top. What do you think? If you like it, I can commit this change to your branch.

image

**Comment by PeteAndersen** [comment]
This page is a little heavy on the SQL queries but it should be OK for now. I think it could be optimized after adding the Family model to bestiary we've discussed in the past.

**Comment by QuatZo** [comment]
> I slightly rearranged to put the legend on top. What do you think? If you like it, I can commit this change to your branch.

It looks great! I think you can also add `<tr>` for headers you mentioned in other comments (if it's not a problem).

> This page is a little heavy on the SQL queries but it should be OK for now. I think it could be optimized after adding the Family model to bestiary we've discussed in the past.

I tried to use as little SQL hits as possible, that's why I decided to use itertools. Right, Family model in Bestiary would optimize Monster preparation.

**Comment by PeteAndersen** [comment]
Thanks! Great work.

**Pull request merged**